File: app/src/main/java/de/euhm/jlt/dialogs/FilterFragment.java
/*
 * @name FilterFragment.java
 * @author hmueller
 * 
 * Licensed under the Apache License, Version 2.0 (the "License")
 * http://www.apache.org/licenses/LICENSE-2.0
 */
package de.euhm.jlt.dialogs;

import android.app.AlertDialog;
import android.app.Dialog;
import android.content.Context;
import android.content.DialogInterface;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.NumberPicker;

import androidx.annotation.NonNull;
import androidx.fragment.app.DialogFragment;

import java.util.Calendar;
import java.util.Locale;

import de.euhm.jlt.R;
import de.euhm.jlt.utils.Constants;

/**
 * Fragment for editing the date filter
 * @author hmueller
 */
public class FilterFragment extends DialogFragment {

	public interface OnFilterFragmentListener {
		void onFinishFilterFragment(int id, int month, int year);
	}

	private OnFilterFragmentListener mListener;
	private static int mMonth = 0;
	private static int mYear = 0;

	public FilterFragment() {
		// create an empty constructor! No args allowed!
	}

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        try {
            mListener = (OnFilterFragmentListener) context;
        } catch (ClassCastException e) {
            throw new ClassCastException(context.toString() + " must implement onFinishFilterFragment");
        }
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        //setHasOptionsMenu(true);
    }

	@Override
	@NonNull
	public Dialog onCreateDialog(Bundle savedInstanceState) {
		if ((mYear == 0) || (mMonth == 0)) {
            throw new NullPointerException("FilterFragment class: setFilter() must be called before onCreateDialog()");
		}

		AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
        // Get the layout inflater
        LayoutInflater inflater = requireActivity().getLayoutInflater();

        // Inflate and set the layout for the dialog
        // Pass null as the parent view because its going in the dialog layout
		final View view = inflater.inflate(R.layout.fragment_edit_filter, null);

		// show the times on the UI
		setPicker(view);

        builder.setView(view)
        	.setTitle(R.string.title_filter)
        	// Add action buttons
        	.setPositiveButton(R.string.button_ok, new DialogInterface.OnClickListener() {
        		@Override
        		public void onClick(DialogInterface dialog, int id) {
        			// handle Ok button
        			int month = getPickerMonth(view);
        			int year= getPickerYear(view);
        			// Return input to the listener/activity
        			mListener.onFinishFilterFragment(Constants.BUTTON_OK, month, year);
        		}
        	})
        	.setNegativeButton(R.string.button_clear, new DialogInterface.OnClickListener() {
        		@Override
        		public void onClick(DialogInterface dialog, int id) {
        			// handle clear filter button
    				// Return input to the listener/activity
    				mListener.onFinishFilterFragment(Constants.BUTTON_CLEAR, 0, 0);
        		}
        	});

        return builder.create();
    }
	
	@Override
	public void onDestroy() {
		super.onDestroy();
		
		// clean up stored references to avoid leaking
		// do not clean static globals, otherwise rotating device crashes 
		mListener = null;
	}

	/**
	 * Set the month of the picker. Use current if 0.
	 * @param month Month to set.
	 */
	public void setPickerMonth(int month) {
		if (month == 0) {
			// if zero, use current month
			month = Calendar.getInstance().get(Calendar.MONTH) + 1;
		}
		mMonth = month;
	}
	
	/**
	 * Set the year of the picker. Use current if 0.
	 * @param year Year to set.
	 */
	public void setPickerYear(int year) {
		if (year == 0) {
			// if zero, use current year
			year = Calendar.getInstance().get(Calendar.YEAR);
		}
		mYear = year;
	}

	/**
	 * Set the data in the picker view (month, year, limits)
	 * 
	 * @param view The current view of the picker
	 */
	private void setPicker(View view) {
		// set month picker
		NumberPicker monthPicker = view.findViewById(R.id.numberPicker_filter_month);
		monthPicker.setMinValue(1);
		monthPicker.setMaxValue(12);
		//monthPicker.setWrapSelectorWheel(false);
		//monthPicker.setFormatter(value -> String.format(Locale.getDefault(), "%02d", value));
		monthPicker.setFormatter(new NumberPicker.Formatter() {
		    @Override
		    public String format(int value) {
		        return String.format(Locale.getDefault(), "%02d", value);
		    }
		});
		monthPicker.setValue(mMonth);

		// set year picker
		NumberPicker yearPicker = view.findViewById(R.id.numberPicker_filter_year);
		yearPicker.setMinValue(1970);
		yearPicker.setMaxValue(2999);
		yearPicker.setWrapSelectorWheel(false);
		yearPicker.setValue(mYear);
	}

	/**
	 * Get the month of the filter picker
	 * 
	 * @param view Current view of the picker
	 * @return Selected month
	 */
	public static int getPickerMonth(View view) {
		// get month picker
		NumberPicker monthPicker = view.findViewById(R.id.numberPicker_filter_month);
		return monthPicker.getValue();
	}

	/**
	 * Get the year of the filter picker
	 * 
	 * @param view Current view of the picker
	 * @return Selected year
	 */
	public static int getPickerYear(View view) {
		// get year picker
		NumberPicker yearPicker = view.findViewById(R.id.numberPicker_filter_year);
		return yearPicker.getValue();
	}
}

File: app/src/main/java/de/euhm/jlt/dialogs/FilterFragment.kt
/**
 * @file FilterFragment.kt
 *
 * Converted from Java to Kotlin by Android Studio Meerkat | 2024.3.1 in March 2025
 *
 * MIT License
 * Copyright (c) 2014-2025 Holger Mueller
 */
package de.euhm.jlt.dialogs

import android.app.AlertDialog
import android.app.Dialog
import android.content.Context
import android.os.Bundle
import android.view.View
import android.widget.NumberPicker
import androidx.fragment.app.DialogFragment
import de.euhm.jlt.R
import de.euhm.jlt.utils.Constants
import java.util.Calendar
import java.util.Locale

/**
 * Fragment for editing the date filter
 * @author hmueller
 */
class FilterFragment : DialogFragment() {
    interface OnFilterFragmentListener {
        fun onFinishFilterFragment(id: Int, month: Int, year: Int)
    }

    private lateinit var mListener: OnFilterFragmentListener
    private var mMonth = 0
    private var mYear = 0

    override fun onAttach(context: Context) {
        super.onAttach(context)
        try {
            mListener = context as OnFilterFragmentListener
        } catch (e: ClassCastException) {
            throw ClassCastException("$context must implement onFinishFilterFragment")
        }
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        if ((mYear == 0) || (mMonth == 0)) {
            throw NullPointerException("FilterFragment class: setFilter() must be called before onCreateDialog()")
        }

        val builder = AlertDialog.Builder(requireActivity())
        // Get the layout inflater
        val inflater = requireActivity().layoutInflater

        // Inflate and set the layout for the dialog
        // Pass null as the parent view because its going in the dialog layout
        val view = inflater.inflate(R.layout.fragment_edit_filter, null)

        // show the times on the UI
        setPicker(view)

        builder.setView(view).setTitle(R.string.title_filter) // Add action buttons
            .setPositiveButton(R.string.button_ok) { _, _ ->
                // handle Ok button
                val month = getPickerMonth(view)
                val year = getPickerYear(view)
                // Return input to the listener/activity
                mListener.onFinishFilterFragment(Constants.BUTTON_OK, month, year)
            }.setNegativeButton(R.string.button_clear) { _, _ ->
                // handle clear filter button
                // Return input to the listener/activity
                mListener.onFinishFilterFragment(Constants.BUTTON_CLEAR, 0, 0)
            }

        return builder.create()
    }

    /**
     * Set the month of the picker. Use current if 0.
     *
     * @param month The month to set (1-12), or 0 to use the current month.
     * @return This [FilterFragment] instance for chaining.
     * @throws IllegalArgumentException if month is outside the range of 0-12.
     */
    fun setPickerMonth(month: Int): FilterFragment {
        require(month in 0..12) { "Month must be between 0 and 12." }
        mMonth = if (month == 0) {
            // Use current month if 0 is provided.
            Calendar.getInstance()[Calendar.MONTH] + 1
        } else {
            month
        }
        return this
    }

    /**
     * Set the year of the picker. Use current if 0.
     *
     * @param year Year to set.
     * @return This [FilterFragment] instance for chaining.
     */
    fun setPickerYear(year: Int): FilterFragment {
        mYear = if (year == 0) {
            // if zero, use current year
            Calendar.getInstance()[Calendar.YEAR]
        } else {
            year
        }
        return this
    }

    /**
     * Set the data in the picker view (month, year, limits)
     *
     * @param view The current view of the picker
     */
    private fun setPicker(view: View) {
        // set month picker
        val monthPicker = view.findViewById<NumberPicker>(R.id.numberPicker_filter_month)
        monthPicker.minValue = 1
        monthPicker.maxValue = 12
        //monthPicker.setWrapSelectorWheel(false);
        monthPicker.setFormatter { value: Int ->
            String.format(Locale.getDefault(), "%02d", value)
        }
        monthPicker.value = mMonth

        // set year picker
        val yearPicker = view.findViewById<NumberPicker>(R.id.numberPicker_filter_year)
        yearPicker.minValue = 1970
        yearPicker.maxValue = 2999
        yearPicker.wrapSelectorWheel = false
        yearPicker.value = mYear
    }

    /**
     * Get the month of the filter picker
     *
     * @param view Current view of the picker
     * @return Selected month
     */
    private fun getPickerMonth(view: View): Int {
        // get month picker
        val monthPicker = view.findViewById<NumberPicker>(R.id.numberPicker_filter_month)
        return monthPicker.value
    }

    /**
     * Get the year of the filter picker
     *
     * @param view Current view of the picker
     * @return Selected year
     */
    private fun getPickerYear(view: View): Int {
        // get year picker
        val yearPicker = view.findViewById<NumberPicker>(R.id.numberPicker_filter_year)
        return yearPicker.value
    }
}